ompi/mca/osc/rdma/osc_rdma_component.c

```diff
@@ -919,56 +910,66 @@ static void ompi_osc_rdma_ensure_local_add_procs (void)
+ * it signficantly simplified osc/rdma's completion. The simplication came in two
+ * areas:
+ *
+ * First, active message RDMA supports remote completion (when a btl's native
+ * RDMA is disabled). Remote completion is required by several key components
+ * of osc/rdma: the usage of cpu atomics to update peer's state, the usage
+ * of local leader to update peer's state, and its fence implementation.
+ * If osc/rdma do not use active message RDMA on alternate btls, it will
+ * have to keep track of each selected btl's support of remote completion.
+ * If any selected btl does not support remote completion, it will have to
+ * disable the usage of cpu atomics, disable the usage of local leader,
+ * and implement a different fence mechanism.
+ *
+ * Second, active message RDMA does not use memory registration explicitly,
+ * therefore using it eliminates the need to store and exchange multiple
+ * memory registrations.
  */
 static int ompi_osc_rdma_query_alternate_btls (ompi_communicator_t *comm, ompi_osc_rdma_module_t *module)
 {
     mca_btl_base_selected_module_t *item;
-    char **btls_to_use = opal_argv_split (ompi_osc_rdma_btl_alternate_names, ',');
     int btls_found = 0;
 
-    btls_to_use = opal_argv_split (ompi_osc_rdma_btl_alternate_names, ',');
-    if (NULL == btls_to_use) {
-        OSC_RDMA_VERBOSE(MCA_BASE_VERBOSE_INFO, "no alternate BTLs requested: %s", ompi_osc_rdma_btl_alternate_names);
-        return OMPI_ERR_UNREACH;
-    }
-
     if (module) {
         module->btls_in_use = 0;
     }
 
     /* rdma and atomics are only supported with BTLs at the moment */
-    for (int i = 0 ; btls_to_use[i] ; ++i) {
-        OSC_RDMA_VERBOSE(MCA_BASE_VERBOSE_INFO, "checking for btl %s", btls_to_use[i]);
-        OPAL_LIST_FOREACH(item, &mca_btl_base_modules_initialized, mca_btl_base_selected_module_t) {
-            if (NULL != item->btl_module->btl_register_mem) { 
-                OSC_RDMA_VERBOSE(MCA_BASE_VERBOSE_INFO, "skipping RDMA btl when searching for alternate BTL");
-                continue;
-            }
-
-            if (0 != strcmp (btls_to_use[i], item->btl_module->btl_component->btl_version.mca_component_name)) {
-                OSC_RDMA_VERBOSE(MCA_BASE_VERBOSE_INFO, "skipping btl %s",
-                                 item->btl_module->btl_component->btl_version.mca_component_name);
-                continue;
-            }
+    OPAL_LIST_FOREACH(item, &mca_btl_base_modules_initialized, mca_btl_base_selected_module_t) {
+        OSC_RDMA_VERBOSE(MCA_BASE_VERBOSE_INFO, "found alternate btl %s",
+                         item->btl_module->btl_component->btl_version.mca_component_name);
 
-            OSC_RDMA_VERBOSE(MCA_BASE_VERBOSE_INFO, "found alternate btl %s", btls_to_use[i]);
+        OSC_RDMA_VERBOSE(MCA_BASE_VERBOSE_INFO, "disabing btl's native support of RDMA and ATOMIC");
+        item->btl_module->btl_flags &= ~(MCA_BTL_FLAGS_RDMA | MCA_BTL_FLAGS_GET | MCA_BTL_FLAGS_PUT | MCA_BTL_FLAGS_ATOMIC_FOPS );
```

Comment:
I'm really not comfortable with this approach; we should not be modifying the BTL module's flags.  Those are owned by the BTL itself.  Why do we need to do this (rather than use our own state for capabilities)?

Reply:
This is necessary because `mca_am_rdma_init` will only fill the gaps, e.g. if a btl does not support put/get/atomics, `mca_am_rdma_init` add the support. `mca_am_rdma_init` will not overwrite btl's native support.

but what osc/rdma want is for an alternate btl to always use AM rdma, do not use its native put/get/atomics. 

Maybe we can add an flag to `mca_am_rdma_init` to force it overwrite a btl's native implementation of rdma/atomics?

Reply:
I think we should go the flags to am_rdma_init route.  I also don't know that we want to overwrite the btl's native implementation; remember that the PML is using the BTL at the same time.

I think we want a proxy btl for the one-sided to use, which is just the underlying BTL and all the AM structures.  So less am_rdma_init() and more btl* btl_base_am_rdma_alloc(base_btl);

Reply:
Thanks for the advice. I will add a function `btl_base_am_rdma_alloc` then

Reply:
@bwbarrett  I tried to allocate a new btl module, but encountered an issue.

The newly created btls are stored in "module->selected_btls`. Later when creating peer, osc/rdma need to pick a btl (from `module->selected_btls`) for each peer. This is currently done by comparing pointer in `module->selected_btls` with pointer of btl registered with bml. Because the newly created btls are not registered with bml, osc/rdma cannot find a btl (and an endpoint) for the peer. (this happens in function `ompi_osc_rdma_peer_btl_endpoint`)

I wonder what is the best way to handle this issue?

Reply:
Yeah, we'd need to go the whole way and import into the BML if we go down that path.  I sent a long email to you and Nathan about some alternatives that might be less costly.  Let's run that to ground and come back to this discussion.